File: dat-maven/src/main/kotlin/me/saro/dat/dat/DatCmsManager.kt
package me.saro.dat.dat

import me.saro.dat.exception.DatErrorCode
import me.saro.dat.exception.DatException
import me.saro.dat.exception.DatResult
import me.saro.dat.exception.DatRetry
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference
import java.util.concurrent.locks.ReentrantReadWriteLock

class DatCmsManager private constructor(
    private val uri: String,
    private var token: String,
    private var version: Long,
    private val manager: DatManager,
    private val client: HttpClient,
    private val scheduler: ScheduledExecutorService?,
    private val requestTimeout: Duration?,
) : AutoCloseable {
    private val lock = ReentrantReadWriteLock()
    private val sync = Runnable { sync() }

    private val lastErrorRef = AtomicReference<DatException?>(DatException(DatErrorCode.CMS_NOT_SYNCED))

    fun getManager() = manager

    fun lastError(): DatException? = lastErrorRef.get()

    fun getVersion(): Long = version

    fun issue(plain: ByteArray, secure: ByteArray): DatResult<String> = manager.issue(plain, secure)

    fun issue(plain: String, secure: String): DatResult<String> = manager.issue(plain, secure)

    fun parse(dat: Dat): DatResult<Payload> = manager.parse(dat)

    fun parse(dat: String?): DatResult<Payload> = manager.parse(dat)

    fun parseWithoutVerifying(dat: Dat): DatResult<Payload> = manager.parseWithoutVerifying(dat)

    fun parseWithoutVerifying(dat: String?): DatResult<Payload> = manager.parseWithoutVerifying(dat)

    fun sync() {
        val error = try {
            syncOrThrow()
            null
        } catch (e: DatException) {
            e
        } catch (e: Exception) {
            DatException(DatErrorCode.CMS_UNKNOWN, "unclassified cms failure", e)
        }
        when {
            error == null -> lastErrorRef.set(null)
            error.retry == DatRetry.STATE -> Unit
            else -> {
                lastErrorRef.set(error)
                log.error("[CRITICAL] DAT CMS SYNC {}: {}", uri, error.message)
            }
        }
    }

    @Throws(DatException::class)
    fun syncOrThrow() {
        if (!lock.writeLock().tryLock()) {
            log.debug("cms sync skipped, previous sync still running: {}", uri)
            throw DatException(DatErrorCode.CMS_SYNC_IN_PROGRESS)
        }
        val newUrl = "$uri?version=$version"
        try {
            val requestBuilder = HttpRequest.newBuilder()
                .uri(URI.create(newUrl))
                .header("Authorization", token)
            requestTimeout?.let { requestBuilder.timeout(it) }
            val request = requestBuilder.build()

            val result = try {
                client.send(request, HttpResponse.BodyHandlers.ofByteArray())
            } catch (e: Exception) {
                throw DatException(DatErrorCode.CMS_UNREACHABLE, "cannot reach $uri", e)
            }

            val status = result.statusCode()
            if (status < 200 || status > 299) {
                throw httpStatusError(status)
            }

            val bytes = result.body()
            if (bytes.isEmpty() || bytes.any { (it.toInt() and 0xff) > 0x7f || it == 0.toByte() }) {
                throw DatException(DatErrorCode.CMS_MALFORMED, "response is not strict ASCII")
            }
            val body = bytes.toString(Charsets.US_ASCII)
            val iof = body.indexOf("\n")
            val versionLine = if (iof >= 0) body.substring(0, iof) else body
            if (versionLine.isEmpty() || versionLine.any { it !in '0'..'9' }) {
                throw DatException(DatErrorCode.CMS_MALFORMED, "version line is not a plain decimal integer")
            }
            val newVersion = versionLine.toLongOrNull()
                ?: throw DatException(DatErrorCode.CMS_MALFORMED, "version line is not a plain decimal integer")
            val newCertificates = if (iof >= 0) body.substring(iof + 1) else ""
            if (newCertificates.isNotEmpty()) {
                if (newVersion < version) {
                    log.warn(
                        "{}: server rolled version back {} -> {}",
                        DatErrorCode.CMS_VERSION_RESET.code, version, newVersion
                    )
                }
                val renew = try {
                    manager.imports(newCertificates, false)
                } catch (e: Exception) {
                    throw DatException(DatErrorCode.CMS_IMPORT_FAILED, "cannot apply received certificates", e)
                }
                version = newVersion
                log.debug("renew {} certificates: {}", renew, newUrl)
            } else {
                log.debug("no new certificate: {}", newUrl)
            }
        } finally {
            lock.writeLock().unlock()
        }
    }

    override fun close() {
        scheduler?.shutdownNow()
        scheduler?.awaitTermination(15, TimeUnit.SECONDS)
        lock.writeLock().lock()
        lock.writeLock().unlock()
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(DatCmsManager::class.java)
        private const val DAT_CMS_API_VERSION = "v1"

        @JvmStatic
        fun builder(): DatCmsManagerBuilder = DatCmsManagerBuilder()

        internal fun httpStatusError(status: Int): DatException = when {
            status == 401 -> DatException(DatErrorCode.CMS_UNAUTHORIZED, "http 401")
            status == 403 -> DatException(DatErrorCode.CMS_FORBIDDEN, "http 403")
            status == 404 -> DatException(DatErrorCode.CMS_ENDPOINT_NOT_FOUND, "http 404")
            status in 500..599 -> DatException(DatErrorCode.CMS_SERVER_ERROR, "http $status")
            else -> DatException(DatErrorCode.CMS_HTTP_STATUS, "http $status")
        }
    }

    class DatCmsManagerBuilder private constructor(
        private var client: HttpClient? = null,
        private var uri: URI = URI.create("http://localhost:8088"),
        private var token: String = "",
        private var verifyOnly: Boolean = false,
        private var intervalSeconds: Long = 60L,
        private var connectTimeoutSeconds: Long = 5L,
        private var requestTimeoutSeconds: Long = 15L,
    ) {
        constructor(): this(
            client = null
        )

        fun uri(uri: String) = this.apply {
            this.uri = try {
                URI.create(uri)
            } catch (e: Exception) {
                throw DatException(DatErrorCode.CONFIG_URI_INVALID, "cannot be parsed as a uri", e)
            }
        }
        fun token(token: String) = this.apply { this.token = token; }
        fun verifyOnly(verifyOnly: Boolean) = this.apply { this.verifyOnly = verifyOnly; }
        fun httpClient(client: HttpClient) = this.apply { this.client = client; }
        fun intervalSeconds(intervalSeconds: Long) = this.apply { this.intervalSeconds = intervalSeconds; }
        fun intervalOff() = this.apply { this.intervalSeconds = 0L; }
        fun connectTimeoutSeconds(seconds: Long) = this.apply { require(seconds >= 0); this.connectTimeoutSeconds = seconds }
        fun requestTimeoutSeconds(seconds: Long) = this.apply { require(seconds >= 0); this.requestTimeoutSeconds = seconds }

        fun build(): DatCmsManager {
            val scheme = this.uri.scheme
            if (scheme != "http" && scheme != "https") {
                throw DatException(DatErrorCode.CONFIG_URI_INVALID, "scheme must be http or https: ${this.uri}")
            }
            if ((this.uri.path?.length ?: 0) > 1) {
                throw DatException(DatErrorCode.CONFIG_URI_INVALID, "must be path-less: ${this.uri}")
            }
            if ((this.uri.query?.length ?: 0) > 0) {
                throw DatException(DatErrorCode.CONFIG_URI_INVALID, "must be query-less: ${this.uri}")
            }
            val path = if (this.verifyOnly) {
                "/$DAT_CMS_API_VERSION/certs/verify-only"
            } else {
                "/$DAT_CMS_API_VERSION/certs"
            }
            val uri = this.uri.toString().trimEnd('/') + path

            val manager = DatManager.newInstance()

            val scheduler: ScheduledExecutorService? = if (intervalSeconds > 0) {
                Executors.newSingleThreadScheduledExecutor { runnable: Runnable ->
                    val thread = Thread(runnable)
                    thread.setDaemon(true)
                    thread.setName("dat-cms-sync-scheduler")
                    thread
                }
            } else {
                null
            }
            val client = client ?: HttpClient.newBuilder().apply {
                if (connectTimeoutSeconds > 0) connectTimeout(Duration.ofSeconds(connectTimeoutSeconds))
                followRedirects(HttpClient.Redirect.NEVER)
            }.build()
            val timeout = requestTimeoutSeconds.takeIf { it > 0 }?.let(Duration::ofSeconds)
            val cms = DatCmsManager(uri, token, 0, manager, client, scheduler, timeout)
            scheduler?.apply {
                scheduleAtFixedRate(cms.sync, intervalSeconds, intervalSeconds, TimeUnit.SECONDS)
            }
            cms.sync()
            return cms
        }
    }
}

File: dat-npm/src/dat.cms.manager.ts
import {DatManager, DatPayload, Dat} from "./index.js";
import {DatError, DatErrorCodes} from "./error.js";

export type Logger = {
    debug: (...args: unknown[]) => void,
    info: (...args: unknown[]) => void,
    warn: (...args: unknown[]) => void,
    error: (...args: unknown[]) => void,
};

export class DatCmsManager {
    private uri: string;
    private token: string;
    private version: number;
    private manager: DatManager;
    private scheduler: any;
    private activeController: AbortController|null = null;
    private stopped: boolean = false;
    private connectTimeoutMs: number;
    private syncTimeoutMs: number;
    private isSyncing: boolean = false;
    private _logger: Logger|any;
    private _lastError: DatError|null = new DatError(DatErrorCodes.CMS_NOT_SYNCED);

    private constructor(
        uri: string,
        token: string,
        version: number,
        manager: DatManager,
        scheduler: any,
        connectTimeoutMs: number,
        syncTimeoutMs: number,
    ) {
        this.uri = uri;
        this.token = token;
        this.version = version;
        this.manager = manager;
        this.scheduler = scheduler;
        this.connectTimeoutMs = connectTimeoutMs;
        this.syncTimeoutMs = syncTimeoutMs;
    }

    getManager(): DatManager {
        return this.manager;
    }

    lastError(): DatError|null {
        return this._lastError;
    }

    getVersion(): number {
        return this.version;
    }

    async issue(plain: Uint8Array | string, secure: Uint8Array | string): Promise<string> {
        return this.manager.issue(plain, secure);
    }

    async parse(dat: Dat | string): Promise<DatPayload> {
        return this.manager.parse(dat);
    }

    async sync(): Promise<void> {
        try {
            await this.syncOrThrow();
            this._lastError = null;
        } catch (e) {
            const err = e instanceof DatError
                ? e : new DatError(DatErrorCodes.CMS_UNKNOWN, "unclassified cms failure", e);
            if (err.retry !== "state") {
                this._lastError = err;
                this._logger.error(`[CRITICAL] DAT CMS SYNC ${this.uri}`, err.code, err);
            }
        }
    }

    async syncOrThrow(): Promise<void> {
        if (this.isSyncing) {
            this._logger.debug(`cms sync skipped, previous sync still running: ${this.uri}`);
            throw new DatError(DatErrorCodes.CMS_SYNC_IN_PROGRESS);
        }

        this.isSyncing = true;
        const newUrl = `${this.uri}?version=${this.version}`;
        let controller: AbortController|null = null;
        let totalTimer: ReturnType<typeof setTimeout>|null = null;
        let connectTimer: ReturnType<typeof setTimeout>|null = null;

        try {
            controller = new AbortController();
            this.activeController = controller;
            totalTimer = this.syncTimeoutMs > 0 ? setTimeout(() => controller!.abort(), this.syncTimeoutMs) : null;
            connectTimer = this.connectTimeoutMs > 0 ? setTimeout(() => controller!.abort(), this.connectTimeoutMs) : null;
            let response: Response;
            try {
                response = await fetch(newUrl, {
                    headers: {
                        'Authorization': this.token
                    },
                    redirect: "manual",
                    signal: controller.signal,
                });
                if (connectTimer) clearTimeout(connectTimer);
            } catch (e) {
                throw new DatError(DatErrorCodes.CMS_UNREACHABLE, `cannot reach ${this.uri}`, e);
            }

            if (!response.ok) {
                throw DatCmsManager.httpStatusError(response.status);
            }

            let bytes: Uint8Array;
            try {
                bytes = new Uint8Array(await response.arrayBuffer());
            } catch (e) {
                throw new DatError(DatErrorCodes.CMS_UNREACHABLE, `cannot read ${this.uri}`, e);
            }
            if (bytes.some(byte => byte > 0x7f)) {
                throw new DatError(DatErrorCodes.CMS_MALFORMED, "response is not strict ASCII");
            }
            const body = new TextDecoder("ascii", {fatal: true}).decode(bytes);
            if (body.length === 0) {
                throw new DatError(DatErrorCodes.CMS_MALFORMED, "response is empty");
            }
            const iof = body.indexOf("\n");
            const versionLine = iof < 0 ? body : body.substring(0, iof);
            if (versionLine.length === 0) {
                throw new DatError(DatErrorCodes.CMS_MALFORMED, "response has no version line");
            }
            if (!/^[0-9]+$/.test(versionLine)) {
                throw new DatError(DatErrorCodes.CMS_MALFORMED, "version line is not a plain decimal integer");
            }
            const newVersion = Number(versionLine);
            if (!Number.isSafeInteger(newVersion)) {
                throw new DatError(DatErrorCodes.CMS_MALFORMED, "version line exceeds the safe integer range");
            }
            const newCertificates = iof < 0 ? "" : body.substring(iof + 1).trim();
            if (newCertificates.length > 0) {
                if (newVersion < this.version) {
                    this._logger.warn(DatErrorCodes.CMS_VERSION_RESET, this.version, newVersion);
                }
                let renew: number;
                try {
                    renew = await this.manager.imports(newCertificates, false);
                } catch (e) {
                    throw new DatError(DatErrorCodes.CMS_IMPORT_FAILED, "cannot apply received certificates", e);
                }
                this.version = newVersion;
                this._logger.debug(`renew ${renew} certificates: ${newUrl}`);
            } else {
                this._logger.debug(`no new certificate: ${newUrl}`);
            }
        } finally {
            if (totalTimer) clearTimeout(totalTimer);
            if (connectTimer) clearTimeout(connectTimer);
            this.activeController = null;
            this.isSyncing = false;
        }
    }

    private static httpStatusError(status: number): DatError {
        switch (status) {
            case 401: return new DatError(DatErrorCodes.CMS_UNAUTHORIZED, "http 401");
            case 403: return new DatError(DatErrorCodes.CMS_FORBIDDEN, "http 403");
            case 404: return new DatError(DatErrorCodes.CMS_ENDPOINT_NOT_FOUND, "http 404");
        }
        if (status >= 500 && status <= 599) {
            return new DatError(DatErrorCodes.CMS_SERVER_ERROR, `http ${status}`);
        }
        return new DatError(DatErrorCodes.CMS_HTTP_STATUS, `http ${status}`);
    }

    stop(): void {
        this.stopped = true;
        this.activeController?.abort();
        if (this.scheduler) {
            clearInterval(this.scheduler);
            this.scheduler = null;
        }
    }

    static builder(): DatCmsManagerBuilder {
        return new DatCmsManagerBuilder();
    }
}

class DatCmsManagerBuilder {
    private _uri: URL = new URL("http://localhost:8088");
    private _token: string = "";
    private _verifyOnly: boolean = false;
    private _intervalSeconds: number = 60;
    private _connectTimeoutSeconds: number = 5;
    private _syncTimeoutSeconds: number = 15;
    private _logger: Logger|any = {
        debug: () => {},
        info: () => {},
        warn: () => {},
        error: () => {}
    };

    uri(uri: string): this {
        try {
            this._uri = new URL(uri);
        } catch (e) {
            throw new DatError(DatErrorCodes.CONFIG_URI_INVALID, "cannot be parsed as a uri", e);
        }
        if (this._uri.protocol !== 'http:' && this._uri.protocol !== 'https:') {
            throw new DatError(DatErrorCodes.CONFIG_URI_INVALID, "scheme must be http or https");
        }
        return this;
    }

    token(token: string): this {
        this._token = token;
        return this;
    }

    verifyOnly(verifyOnly: boolean): this {
        this._verifyOnly = verifyOnly;
        return this;
    }

    logger(logger: Logger|any): this {
        this._logger = logger;
        return this;
    }

    intervalSeconds(intervalSeconds: number): this {
        this._intervalSeconds = intervalSeconds;
        return this;
    }

    intervalOff(): this {
        this._intervalSeconds = 0;
        return this;
    }

    connectTimeoutSeconds(seconds: number): this { this._connectTimeoutSeconds = seconds; return this; }
    syncTimeoutSeconds(seconds: number): this { this._syncTimeoutSeconds = seconds; return this; }

    async build(): Promise<DatCmsManager> {
        if (this._uri.pathname.length > 1) {
            throw new DatError(DatErrorCodes.CONFIG_URI_INVALID, `must be path-less: ${this._uri}`);
        }
        if (this._uri.search.length > 0) {
            throw new DatError(DatErrorCodes.CONFIG_URI_INVALID, `must be query-less: ${this._uri}`);
        }

        const path = this._verifyOnly ? "/v1/certs/verify-only" : "/v1/certs";
        const uri = `${this._uri.protocol}//${this._uri.host}${path}`;

        const manager = new DatManager();
        let scheduler: any = null;

        const cms = new (DatCmsManager as any)(uri, this._token, 0, manager, null,
            this._connectTimeoutSeconds * 1000, this._syncTimeoutSeconds * 1000);
        cms._logger = this._logger;
        
        await cms.sync();

        if (this._intervalSeconds > 0) {
            scheduler = setInterval(() => {
                cms.sync();
            }, this._intervalSeconds * 1000);
            (cms as any).scheduler = scheduler;
        }

        return cms;
    }
}
